[Logan] Performance improvements for the diffing algorithm.

ian97531 opened the pull request:
With this change, I'm seeing 50x - 100x improvements in overall diffing performance. Documents that previously took ~26seconds now take ~300ms.

I kept the basic algorithm the same, but changed it from recursive to iterative, and added a bunch of little optimizations.

loganfsmyth commented on js/htmldiff.js:
```diff
@@ -109,12 +123,19 @@
-  function Match(start_in_before, start_in_after, length){
-    this.start_in_before = start_in_before;
-    this.start_in_after = start_in_after;
+  function Match(start_in_before, start_in_after, length, segment){
```
Needs JSDocs for `segment`.

ian97531 replied:
Resolved.

loganfsmyth commented on js/htmldiff.js:
```diff
@@ -109,12 +123,19 @@
+    this.segment_end_in_after = (this.segment_start_in_after + this.length) - 1;
+
+    this.end_in_before = segment.before_index + this.segment_end_in_before;
+    this.end_in_after = segment.after_index + this.segment_end_in_after;
```
These were hard for me to process. Think it'd be clearer as

    this.end_in_before = this.start_in_before + this.length - 1;
    this.end_in_after = this. start_in_after + this.length - 1;

loganfsmyth commented on js/htmldiff.js:
```diff
@@ -437,7 +680,8 @@
       var op = operations[i];
 
-      if ((is_single_whitespace(op) && last_op.action === 'replace') || (op.action === 'replace' && last_op.action === 'replace')){
+      if ((is_single_whitespace(op) && last_op.action === 'replace') || 
```
FYI, looks like this PR has a ton of trailing spaces on lines like this.

loganfsmyth commented:
Since we're splitting this deploy, I landed my other PR with updates, so this has a conflict now. Should be easy to resolve though.

ian97531 commented:
@loganfsmyth I think I've addressed all of your comments. I re-ran the tests and they look good. I also verified by diffing some real content and it looks good as well.

loganfsmyth commented:
LGTM
